File: templating/rule_contributions/src/main/python/jsonteng/rules/ipv4_host_ip_rule.py
import ipaddress

from jsonteng.exception import TemplateEngineException
from jsonteng.rules.rule_base import RuleBase


class Ipv4HostIpRule(RuleBase):
    """
    Network rule for returning network address, netmask and gateway.
    Gateway is assumed to be the lowest network address.
    """
    name = "ipv4-host-ip"

    def __init__(self, rule_resolver):
        """
        Construct this rule.
        :param rule_resolver: Rule resolver
        :type rule_resolver: 'RuleResolver'
        """
        super().__init__(rule_resolver)
        self._element_resolver = rule_resolver.get_element_resolver()

    def process(self, rule_tokens, binding_data_list):
        """
        Process this rule.
        :param binding_data_list: Binding data used during the processing.
        :type binding_data_list: 'list'
        :param rule_tokens: Rule arguments.
        :type rule_tokens: 'list'
        :return: JSON object
        :rtype: JSON object
        """
        token_count = len(rule_tokens)
        if token_count < 2:
            raise TemplateEngineException(
                "Rule \"{}\" needs at least 2 parameter."
                " Parameters given {}".format(Ipv4HostIpRule.name, rule_tokens))
        network = ipaddress.ip_network(self._element_resolver.resolve(
            rule_tokens[0], binding_data_list))
        index = int(self._element_resolver.resolve(
            rule_tokens[1], binding_data_list))
        return str(ipaddress.ip_address(
            int(ipaddress.ip_address(network.network_address)) + index))

File: templating/rule_contributions/src/main/java/com/vmware/jsonteng/ext/rules/net/Ipv4HostIpRule.java
package com.vmware.jsonteng.ext.rules.net;

import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.nio.ByteBuffer;
import java.util.List;
import java.util.Map;

import com.vmware.jsonteng.ElementResolver;
import com.vmware.jsonteng.RuleResolver;
import com.vmware.jsonteng.TemplateEngineException;
import com.vmware.jsonteng.rules.RuleBase;

public class Ipv4HostIpRule extends RuleBase {
    public static final String name = "ipv4-host-ip";

    private final ElementResolver elementResolver;

    public Ipv4HostIpRule(RuleResolver ruleResolver) {
        super(ruleResolver);
        elementResolver = ruleResolver.getElementResolver();
    }

    @Override
    public Object process(List<?> ruleTokens, List<Map<String, ?>> bindingDataList) throws TemplateEngineException {
        int tokenCount = ruleTokens.size();
        if (tokenCount < 2) {
            throw new TemplateEngineException(
                    String.format("Rule \"%s\" needs at least 2 parameters. Parameters given %s", name, ruleTokens));
        }
        String resolvedNetwork = (String) this.elementResolver.resolve(
                ruleTokens.get(0), bindingDataList);
        String[] parts = resolvedNetwork.split("/");
        Inet4Address networkAddress;
        try {
            networkAddress = (Inet4Address) InetAddress.getByName(parts[0]);
        } catch (UnknownHostException e) {
            throw new TemplateEngineException("Unable to get the network address " + parts[0], e);
        }
        ByteBuffer fourBytes = ByteBuffer.wrap(new byte[4]);
        fourBytes.put(networkAddress.getAddress());
        fourBytes.rewind();
        int network = fourBytes.getInt();
        int index = Integer.parseInt(this.elementResolver.resolve(
                ruleTokens.get(1), bindingDataList).toString());
        int prefix = Integer.parseInt(parts[1]);
        if (index == 0 || index >= (1 << (32 - prefix))) {
            throw new TemplateEngineException(index + " is out of ip range 2^" + prefix);
        }
        int ipAddressInt = network + index;
        fourBytes.rewind();
        fourBytes.putInt(ipAddressInt);
        Inet4Address ipAddress;
        try {
            ipAddress = (Inet4Address) InetAddress.getByAddress(fourBytes.array());
        } catch (UnknownHostException e) {
            throw new TemplateEngineException("Unabled to get ip address " + fourBytes, e);
        }
        return ipAddress.getHostAddress();
    }
}
